aws/cloudformation.yml

```diff
@@ -0,0 +1,699 @@
+                echo "$?" >> /var/tmp/bootstrap.txt
+
+                echo "signal trigger" >> /var/tmp/bootstrap.txt
+                /usr/local/bin/cfn-signal --success true --region ${AWS::Region} --stack ${AWS::StackName} --resource EC2InstancePrimary
+                echo "$?" >> /var/tmp/bootstrap.txt
+                #Warning: sudo cfn-signal seems to freeze/break the userdata.
+
+                echo "waiting for stack-create-complete" >> /var/tmp/bootstrap.txt
+                aws cloudformation wait stack-create-complete --stack-name ${AWS::StackName} --region ${AWS::Region}
+                echo "$?" >> /var/tmp/bootstrap.txt
+                echo "CFN Stack completed" >> /var/tmp/bootstrap.txt
+                echo "-----" >> /var/tmp/bootstrap.txt
+
+                echo "script1" >> /var/tmp/bootstrap.txt
+                bash -ex /var/tmp/stage1.sh  # Mastodon User/Docker/Docker-Compose
+                echo "script 1 result $?" >> /var/tmp/bootstrap.txt
+                echo "-----" >> /var/tmp/bootstrap.txt
+
+                echo "script2" >> /var/tmp/bootstrap.txt
+                sudo -u mastodon bash -ex /var/tmp/stage2.sh  # Mastodon
+                echo "script 2 result $?" >> /var/tmp/bootstrap.txt
+                echo "-----" >> /var/tmp/bootstrap.txt
+
+                echo "script3" >> /var/tmp/bootstrap.txt
+                bash -ex /var/tmp/stage3.sh  # Nginx and Certbot
+                echo "script 3 result $?" >> /var/tmp/bootstrap.txt
+                echo "-----" >> /var/tmp/bootstrap.txt
+
+                echo "script4" >> /var/tmp/bootstrap.txt
+                sudo -u mastodon bash -ex /var/tmp/stage4.sh  # Crontab/Maintenance Jobs
+                echo "script 2 result $?" >> /var/tmp/bootstrap.txt
+                echo "-----" >> /var/tmp/bootstrap.txt
+
+                echo "Final reboot" >> /var/tmp/bootstrap.txt
+                systemctl reboot
+
+        create_files:
+          files: 
+            /var/tmp/stage1.sh:
+              content: !Sub |
+                #!/bin/bash -ex
+
+                echo "Create mastodon group" >> /var/tmp/bootstrap.txt
+                addgroup -gid 991 mastodon
+                echo "$?" >> /var/tmp/bootstrap.txt
+
+                echo "Create mastodon user" >> /var/tmp/bootstrap.txt
+                adduser --quiet --disabled-password --gid 991 --uid 991 mastodon --gecos GECOS
+                echo "$?" >> /var/tmp/bootstrap.txt
+
+                echo "Grant mastodon sudo" >> /var/tmp/bootstrap.txt
+                usermod -a -G sudo mastodon
+                echo "$?" >> /var/tmp/bootstrap.txt
+
+                echo "change mastodon password" >> /var/tmp/bootstrap.txt
+                echo "mastodon:${MastodonLocalUserPassword}" | sudo chpasswd
+                echo "$?" >> /var/tmp/bootstrap.txt
+
+                echo "Install Docker" >> /var/tmp/bootstrap.txt
```

Comment:
If we're doing a CF template, then it might be a good idea to avoid using Docker. The stack isn't particularly heavy; adding Docker to the equation is a little bit lazy

Reply:
1) I don't think Docker is lazy, but I also think it's worth it for the upgrade (and fallback) path.

* git fetch
* git stash
* git checkout {newversion}
* git pop
* edit docker-compose.yml image to {newversion}
* docker-compose build
* db:migrate if required
* docker-compose up -d

That's a pretty clean upgrade path?.... I can be convinced to remove all of the docker for the "performance" boot.

2) ECS is not a good fit for my template... and while I think Mastodon could easily be turned into a ECS task.... it's not "microservice" enough atm.

ECS also would assume you're using an ALB in front of your ECS tasks and that would increase the cost.

Reply:
The front end doesn't need to incur downtime if you skip using Docker. But for a v1 CFT, Docker can probably stay